File: Mbs/Trading/Data/Generators/Sawtooth/SawtoothParameters.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;

namespace Mbs.Trading.Data.Generators.Sawtooth
{
    /// <summary>
    /// The input parameters for the sawtooth impulse generator.
    /// </summary>
    public class SawtoothParameters : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the amplitude of the sawtooth impulse, should be positive.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.SawtoothAmplitude)]
        [Range(0.0, double.MaxValue, ErrorMessage = ErrorMessages.FieldMustBePositive)]
        public double Amplitude { get; set; } = DefaultParameterValues.SawtoothAmplitude;

        /// <summary>
        /// Gets or sets the value corresponding to the minimum of the sawtooth impulse, should be positive.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.SawtoothMinimalValue)]
        [Range(0.0, double.MaxValue, ErrorMessage = ErrorMessages.FieldMustBePositive)]
        public double MinimalValue { get; set; } = DefaultParameterValues.SawtoothMinimalValue;

        /// <summary>
        /// Gets or sets a value indicating whether the sawtooth impulse is reflected horizontally to form a triangle shape.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.SawtoothIsBiDirectional)]
        public bool IsBiDirectional { get; set; } = DefaultParameterValues.SawtoothIsBiDirectional;

        /// <summary>
        /// Gets or sets the sawtooth shape.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.SawtoothShape)]
        public SawtoothShape Shape { get; set; } = DefaultParameterValues.SawtoothShape;

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }
}

File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/data/generators/sawtooth/sawtooth-parameters.ts
import { SawtoothShape } from './sawtooth-shape.enum';
import { amplitudeName, minimalValueName, isBiDirectionalName, sawtoothShapeName, objectName } from '../constants';

/** The input parameters for the sawtooth impulse generator. */
export class SawtoothParameters {
  private static readonly defaultAmplitude: number = 100;
  private static readonly defaultMinimalValue: number = 10;
  private static readonly defaultIsBiDirectional: boolean = false;
  private static readonly defaultSawtoothShape: SawtoothShape = SawtoothShape.Linear;

  /** The amplitude of the sawtooth impulse, should be positive. */
  amplitude: number = SawtoothParameters.defaultAmplitude;

  /** The minimum of the sawtooth impulse, should be positive. */
  minimalValue: number = SawtoothParameters.defaultMinimalValue;

  /** Whether the sawtooth impulse is reflected horizontally to form a triangle shape. */
  isBiDirectional: boolean = SawtoothParameters.defaultIsBiDirectional;

  /** The sawtooth shape. */
  sawtoothShape: SawtoothShape = SawtoothParameters.defaultSawtoothShape;

  constructor(data?: SawtoothParameters) {
    if (data) {
      for (const property in data) {
        if (data.hasOwnProperty(property)) {
          (<any>this)[property] = (<any>data)[property];
        }
      }
    }
  }

  toJSON(data?: any) {
    data = typeof data === objectName ? data : {};
    data[amplitudeName] = this.amplitude;
    data[minimalValueName] = this.minimalValue;
    data[isBiDirectionalName] = this.isBiDirectional;
    data[sawtoothShapeName] = this.sawtoothShape;
    return data;
  }
}
